compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1285,19 +1305,21 @@ convertArg env = \case
     Type t -> TyArg <$> convertType env t
     e -> TmArg <$> convertExpr env e
 
-withTyArg :: Env -> (LF.TypeVarName, LF.Kind) -> [LArg Var] -> (LF.Type -> [LArg Var] -> ConvertM (LF.Expr, [LArg Var])) -> ConvertM (LF.Expr, [LArg Var])
+withTyArg :: Env -> LF.Kind -> [LArg Var] -> (LF.Type -> [LArg Var] -> ConvertM (LF.Expr, [LArg Var])) -> ConvertM (LF.Expr, [LArg Var])
 withTyArg env _ (LType t:args) cont = do
     t <- convertType env t
     cont t args
-withTyArg env (v, k) args cont = do
+withTyArg env k args cont = do
+    v <- freshTyVar
     (x, args) <- cont (TVar v) args
     pure (ETyLam (v, k) x, args)
```

Comment:
`withTyArg` should really add the type variable to the Env, and pass a new env' to its descendants, like `envBindTypeVar` does.